pygac/l1c_factory.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2014, 2019 Pygac Developers
+
+# Author(s):
+
+#   Carlos Horn <carlos.horn@external.eumetsat.int>
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Factory for l1c files.
+
+"""
+import datetime
+import logging
+
+from pygac.gac_klm import GACKLMReader
+from pygac.gac_pod import GACPODReader
+from pygac.lac_klm import LACKLMReader
+from pygac.lac_pod import LACPODReader
+from pygac.utils import file_opener
+
+
+LOG = logging.getLogger(__name__)
+
+
+class L1cBuilder(object):
+    def __init__(self, Reader):
+        self.Reader = Reader
+
+    def __call__(self, filename, start_line, end_line, fileobj=None):
+        tic = datetime.datetime.now()
+        LOG.info("Building file: %s", str(filename))
+        reader = self.Reader.fromfile(filename, fileobj=fileobj)
+        reader.save(start_line, end_line)
+        LOG.info("Processing took: %s", str(datetime.datetime.now() - tic))
+
+
+class L1cFactory(object):
```

Comment:
I don't think this needs to be a class, it would be clearer if it was a function returning the reader to use on the given file. It would be useful not only for doing the entire processing as you do in `__call__` but also for external libs that just want to know what reader to use for the file.
Down the line, in gac_run, you can then just do
```python
reader = find_out_what_reader_can_read_this_file(filename)
...
reader.read_and_save(...)
```
which IMO makes it clearer than the `L1cFactory` usage.

Reply:
In which module would you put this function, or should it have its own one?

Reply:
Good question. I suppose `reader.py` would be a good candidate, it that doesn't create circular imports. Otherwise, do we already have a `utils` module ?